srv/modules/runners/mgr_orch.py

```diff
@@ -0,0 +1,172 @@
+
+
+def _sanitize_devices(devices):
+    """
+    Takes a list of devices and returns only the information necessary for
+    the ceph-mgr orchestrator's InventoryDevice class.
+    """
+
+    # Shamelessly stolen from populate.py, to try to prefer /dev/disk/by-id
+    # paths over /dev/sd[a-z]+.  It would probably be better to lean on the
+    # logic in srv/salt/_modules/cephdisks.py:device_() to get the most
+    # preferred path, but calling that for each individual disk is going to
+    # take a very long time!
+    def _device(drive):
+        device = drive['Device File']
+        if 'Device Files' in drive:
+            for path in drive['Device Files'].split(', '):
+                if 'by-id' in path:
+                    device = path
+                    break
+        return device
+
+    # Attempt to figure out if the drive is hdd, ssd or nvme
+    # TODO: this looks right, but is completely untested for ssd and nvme!
+    def _type(drive):
+        type = "hdd"
+        if drive["rotational"] == "0":
+            type = "ssd"
+        if drive["Driver"] == "nvme":
+            type = "nvme"
+        return type
+
+    sanitized = []
+    for drive in devices:
+        sanitized.append({
+            "blank": drive["blank"],
+            "type": _type(drive),
+            "id": _device(drive),
+            "size": int(drive["Bytes"]),
+            "extended": None,
+            "metadata_space_free": None     # TODO: calculate this
+        })
+    return sanitized
+
+
+def get_inventory(nodes=[], roles=[]):
+    """
+    Return an inventory of all devices on all nodes, optionally limited
+    to nodes with particular hostnames or roles.  Criteria are OR'd, e.g.
+    if multiple roles are passed in, this will inventory all nodes assigned
+    to _any_ of the listed roles.  Likewise if multiple nodes are passed,
+    all those nodes will be inventoried.
+
+    Examples:
+
+    salt-run mgr_orch.get_inventory roles=mon
+    salt-run mgr_orch.get_inventory roles=[mgr,mds]
+    salt-run mgr_orch.get_inventory nodes=data1.example.com
+    salt-run mgr_orch.get_inventory nodes=[data1.example.com,data5.example.com]
+    """
+
+    # The cluster is always named 'ceph'
+    search = "I@cluster:ceph"
+
+    # Syntactic sugar (allows passing nodes=foo or roles=bar for single items,
+    # rather than having to add the square brackets around one-item lists)
+    if not isinstance(nodes, list):
+        nodes = [nodes]
+    if not isinstance(roles, list):
+        roles = [roles]
+
+    criteria = nodes
+    for role in roles:
+        criteria.append("I@roles:{}".format(role))
+    if criteria:
+        search = "{} and ( {} )".format(search, " or ".join(criteria))
+
+    # When search matches no minions, salt prints to stdout.  Suppress stdout.
+    _stdout = sys.stdout
+    sys.stdout = open(os.devnull, 'w')
+
+    local = salt.client.LocalClient()
+    _minions = local.cmd(search, 'cephdisks.list', [], tgt_type="compound")
```

Comment:
This can be replaced with the ceph-volume inventory function soon

Reply:
I wondered about that - is there going to be a new runner that wraps ceph-volume, or are we just going to change cephdisks.list to use ceph-volume internally, given cephdisks.list is already used in various other places?

Reply:
> are we just going to change cephdisks.list to use ceph-volume internally, given cephdisks.list is already used in various other places?

This probably makes the most sense.